test/utils/floating_point_utils.hpp

```diff
@@ -0,0 +1,89 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace fizzy::test
+{
+/// Simple implementation of C++20's std::bit_cast.
+template <typename DstT, typename SrcT>
+DstT bit_cast(SrcT x) noexcept
+{
+    DstT z;
+    static_assert(sizeof(x) == sizeof(z));
+    __builtin_memcpy(&z, &x, sizeof(x));
+    return z;
+}
+
+/// A wrapper for floating-point types with inspection/construction/comparison utilities.
+template <typename T>
+struct FP
+{
+    static_assert(std::is_same_v<T, float> || std::is_same_v<T, double>);
+
+    /// Shortcut to numeric_limits.
+    using Limits = std::numeric_limits<T>;
+    static_assert(Limits::is_iec559);
+
+    /// The unsigned integer type matching the size of this floating-point type.
+    using UintType = std::conditional_t<std::is_same_v<T, float>, uint32_t, uint64_t>;
+    static_assert(sizeof(T) == sizeof(UintType));
+
+    /// The number of mantissa bits in the binary representation.
+    static constexpr auto num_mantissa_bits = Limits::digits - 1;
+
+    /// The binary mask of the mantissa part of the binary representation.
+    static constexpr auto mantissa_mask = (UintType{1} << num_mantissa_bits) - 1;
+
+    /// The number of exponent bits (E) in the binary representation.
+    /// Produced out of Limits::max_exponent which is 2^
+    static constexpr auto num_exponent_bits = 2 * Limits::max_exponent;
+
+    /// The binary mask of the exponent part of the binary representation, i.e. 2^E - 1.
+    /// This is also exponent value (all exponent bits set) for NaNs.
+    /// Produced from Limits::max_exponent which is 2^(E-1).
+    static constexpr auto exponent_mask = 2 * UintType{Limits::max_exponent} - 1;
```

Comment:
did you want to use `num_exponent_bits` here?

Reply:
No, it was suppose to be the mask, but in the end I changed that to `nan_exponent` as the mask would be at the wrong position.